File: Core/Kernel/ChoMAWrapper.m
#import "ChoMAWrapper.h"
#import "MachO.h"
#import "CSBlob.h"

@implementation ChoMAWrapper

+ (NSData *)extractCDHash:(NSString *)machOPath {
    if (![[NSFileManager defaultManager] isReadableFileAtPath:machOPath]) return nil;

    struct macho *macho = macho_init_for_writing(machOPath.UTF8String);
    if (!macho) return nil;

    CS_SuperBlob *superblob = macho_read_code_signature(macho);
    if (!superblob) { macho_free(macho); return nil; }

    struct csd_superblob *decoded = csd_superblob_decode(superblob);
    free(superblob);
    if (!decoded) { macho_free(macho); return nil; }

    uint8_t hash[20];
    int32_t hashType = 0;
    int ret = csd_superblob_calculate_best_cdhash(decoded, hash, &hashType);
    csd_superblob_free(decoded);
    macho_free(macho);

    if (ret != 0) return nil;
    return [NSData dataWithBytes:hash length:20];
}

@end

File: Vendored/ChOma/ChoMAWrapper.swift
import Foundation

internal final class ChoMAWrapper {
    enum ChOmaError: LocalizedError {
        case cannotReadBinary(String)
        case entitlementsFailed
        case cannotWriteBinary(String)
        case cdhashFailed

        var errorDescription: String? {
            switch self {
            case .cannotReadBinary(let path): return "Cannot read Mach-O at: \(path)"
            case .entitlementsFailed: return "Failed to create or apply entitlements"
            case .cannotWriteBinary(let path): return "Cannot write modified binary to: \(path)"
            case .cdhashFailed: return "CDHash extraction failed"
            }
        }
    }

    func applyEntitlements(to machOPath: String, xml entitlementsXML: String) throws {
        guard FileManager.default.isReadableFile(atPath: machOPath) else {
            throw ChOmaError.cannotReadBinary(machOPath)
        }

        let tmpDir = FileManager.default.temporaryDirectory
        let tmpFile = tmpDir.appendingPathComponent("entitlements_\(UUID().uuidString).xml")
        try entitlementsXML.write(to: tmpFile, atomically: true, encoding: .utf8)
        defer { try? FileManager.default.removeItem(at: tmpFile) }

        let result = choma_replace_entitlements(machOPath, tmpFile.path)
        if result != 0 {
            throw ChOmaError.entitlementsFailed
        }
    }

    static func extractCDHash(_ machOPath: String) -> Data? {
        guard FileManager.default.isReadableFile(atPath: machOPath) else { return nil }

        guard let macho = macho_init_for_writing(machOPath) else { return nil }
        defer { macho_free(macho) }

        guard let superblob = macho_read_code_signature(macho) else { return nil }
        defer { free(superblob) }

        guard let decoded = csd_superblob_decode(superblob) else { return nil }
        defer { csd_superblob_free(decoded) }

        var cdhash = Data(count: 20)
        var cdhashType: Int32 = 0
        let ret = cdhash.withUnsafeMutableBytes { buf in
            csd_superblob_calculate_best_cdhash(decoded, buf.baseAddress, &cdhashType)
        }
        guard ret == 0 else { return nil }
        return cdhash
    }
}
